### DISCUSSION TITLE
Retrieve material property from LiDAR ray cast for better intensity modelling.

### DISCUSSION THREAD

**Discussion by fgoudreault** [carla-simulator/carla#5598] [Q&A] [answered]
Hello all! I am interested in computing more realistic LiDAR intensity values for modelling. I wondered if there was anybody who knows a bit the Unreal Engine to help me down to get this (I am quite noob in c++ an Unreal Engine API tbh).
I tried modifying the code in RayCastSemanticLidar.cpp in order to do so. In the ComputeRawDetection function where one of the argument is the FHitResult object I tried:
```
auto comp = HitInfo.GetComponent();
int32 sectionIndex = 0;
int32 faceIndex = HitInfo.FaceIndex;
UMaterialInterface materialInterface = comp->GetMaterialFromCollisionFaceIndex(faceIndex, sectionIndex);
```
but it always segfaults for some reason upon execution or returns a nullptr...

I also tried:

```
UMaterial * material = comp->GetMaterial(0)->GetMaterial();
FScalarMaterialInput metallic = materialInterface->GetMaterial()->Metallic;
```

but to no avail...
If someone has any idea and would like to help, fell free to DM me! (this discussion also [posted on the carla discord server](https://discordapp.com/channels/444206285647380482/445494080076513280/998652315441561651)).

**Comment by balaji-ch** [comment]
Did you solve it? Even I need this.

**Reply by fgoudreault** [reply, marked as answer]
I found a workaround (sorry I haven't marked this post as fixed). See [my reply to the corresponding issue](https://github.com/carla-simulator/carla/issues/3826#issuecomment-1573689763).

Basically I used custom cameras to get the material properties of the environment (4 cameras of 90 degrees FOV each to get full 360 degrees) and projected LiDAR point clouds onto them to retrieve the values I wanted (i.e.: metallicity, roughness, diffuse coefficients, etc.).
